aws/resource_aws_ssm_resource_data_sync_test.go

```diff
@@ -0,0 +1,130 @@
+package aws
+
+import (
+	"fmt"
+	"log"
+	"testing"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/ssm"
+	"github.com/hashicorp/terraform/helper/acctest"
+	"github.com/hashicorp/terraform/helper/resource"
+	"github.com/hashicorp/terraform/terraform"
+)
+
+func TestAccAWSSsmResourceDataSync(t *testing.T) {
+	rInt := acctest.RandInt()
+	resource.Test(t, resource.TestCase{
+		PreCheck:     func() { testAccPreCheck(t) },
+		Providers:    testAccProviders,
+		CheckDestroy: testAccCheckAWSSsmResourceDataSyncDestroy,
+		Steps: []resource.TestStep{
+			{
+				Config: testAccSsmResourceDataSyncConfig(rInt),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckAWSSsmResourceDataSyncExists("aws_ssm_resource_data_sync.foo"),
+				),
+			},
+		},
+	})
+}
+
+func testAccCheckAWSSsmResourceDataSyncDestroy(s *terraform.State) error {
+	conn := testAccProvider.Meta().(*AWSClient).ssmconn
+
+	for _, rs := range s.RootModule().Resources {
+		if rs.Type != "aws_ssm_resource_data_sync" {
+			continue
+		}
+		nextToken := ""
+		found := false
+		for {
+			input := &ssm.ListResourceDataSyncInput{}
+			if nextToken != "" {
+				input.NextToken = aws.String(nextToken)
+			}
+			resp, err := conn.ListResourceDataSync(input)
+			if err != nil {
+				return err
+			}
+			for _, v := range resp.ResourceDataSyncItems {
+				if *v.SyncName == rs.Primary.Attributes["name"] {
+					found = true
+				}
+			}
+			if resp.NextToken != nil {
+				nextToken = *resp.NextToken
+			}
+			if found || nextToken == "" {
+				break
+			}
+		}
```

Comment:
Once you have decoupled the logic for finding the right data sync, it can also be reused here 😉